---
title: "evaluating haplotype clustering"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
---

# Statistics of long reads

Long reads are clustered by haplotypes based on their edit distance with heterezygous bubble kmers from haplotype 1 and 2.


```{r}
library(dplyr)
library(data.table) %>% invisible()
library(ggplot2) %>% invisible()
library(gridExtra) %>% invisible()
hdist = fread('/home/maryam/server_mount/SaaRclust/pipeline/evaluation/long_read_phase_evaluation/iteration2_haplo_edit_dist_k63_a3_l23_kminimap15_w1_f0.1_z500_HG00733.data')  %>% invisible()
hdist.filt <- hdist[dist_h0!='None' & dist_h1!='None' & pred_type!='None']  %>% invisible()

```

## Distribution of d1/(d1+d2) these different types of long reads:

```{r}
density.plt <- ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1), fill=pred_type))+geom_density(alpha=0.2)
density.plt

frequency.plt <- ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1), fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
frequency.plt
#suppressWarnings(marrangeGrob(list(density.plt, frequency.plt), nrow=2, ncol=1))

ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1)))+geom_density(aes(y=..count..), alpha=0.2)

ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1)))+geom_density(aes(y=..count..), alpha=0.2)+facet_wrap(~pred_type)

ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1)))+geom_density(alpha=0.2)+facet_wrap(~pred_type)
```

## Distribution of num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles) these different types of long reads:

```{r}
density.plt <- ggplot(hdist.filt, aes(num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles), fill=pred_type))+geom_density(alpha=0.2)
density.plt

frequency.plt <- ggplot(hdist.filt, aes(num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles), fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
frequency.plt
#suppressWarnings(marrangeGrob(list(density.plt, frequency.plt), nrow=2, ncol=1))

ggplot(hdist.filt, aes(num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles)))+geom_density(aes(y=..count..), alpha=0.2)

ggplot(hdist.filt, aes(num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles)))+geom_density(aes(y=..count..), alpha=0.2)+facet_wrap(~pred_type)

ggplot(hdist.filt, aes(num_h0_bubbles/(num_h0_bubbles+num_h1_bubbles)))+geom_density(alpha=0.2)+facet_wrap(~pred_type)
```



## Distribution of number of aligned bubbles for different types of long reads:

```{r}
ggplot(hdist.filt[num_aln_bubbles<100], aes(num_aln_bubbles, fill=pred_type))+geom_histogram(bins=100, alpha=0.5)

ggplot(hdist.filt[num_aln_bubbles<100], aes(num_aln_bubbles, color="blue"))+geom_histogram(aes(y=..density..), bins=100, alpha=0.8)+facet_wrap(~pred_type)

ggplot(hdist.filt[num_aln_bubbles<100], aes(num_aln_bubbles, color="blue"))+geom_histogram(bins=100, alpha=0.8)+facet_wrap(~pred_type)
```

## Distribution of average bubbles kmer coverage in different groups of long reads:

```{r}
#ggplot(hdist.filt[num_aln_bubbles<100], aes(avg_km, fill=pred_type))+geom_histogram(aes(y=..density..), alpha=0.2)
```

# Stattistics of SNV bubbles

Now we plot the (d1, h2) edit distance with the first allele of bubbles with haplotype1 and haplotype2 kmers, respectively. Bubbles are grouped by their prediction type in {"not_chrom_clust", "haploclust_false_pos", "false_haplo_clust",  "true_haplo_clust"      "haplo_clust_false_neg" "not_haplo_clust"}.

```{r}
# hdist = fread('/home/maryam/server_mount/SaaRclust/pipeline/evaluation/bubble_phase_evaluation/iteration2_haplo_edit_dist_k63_a3_l23_kminimap15_w1_f0.1_z500_HG00733.data')
hdist.filt <- hdist[dist_h0!='None' & dist_h1!='None' & pred_type!='None']

#ggplot(hdist.filt, aes(x=dist_h1, y=dist_h2, color=pred_type))+geom_point()
```

## The density plots of each of these bubble groups for d1/(d1+d2):

```{r}
density.plt <- ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1), fill=pred_type))+geom_density(alpha=0.2)
frequency.plt <- ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1), fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
suppressWarnings(marrangeGrob(list(density.plt, frequency.plt), nrow=2, ncol=1))

ggplot(hdist.filt, aes(dist_h0/(dist_h0+dist_h1)))+geom_density(alpha=0.2)+facet_wrap(~pred_type)
```
## The density plots of each of these bubble groups for num_al0_h0_long_reads/(num_al0_h0_long_reads+num_al0_h1_long_reads):

```{r}
density.plt <- ggplot(hdist.filt, aes(num_al0_h0_long_reads/(num_al0_h0_long_reads+num_al0_h1_long_reads), fill=pred_type))+geom_density(alpha=0.2)
frequency.plt <- ggplot(hdist.filt, aes(num_al0_h0_long_reads/(num_al0_h0_long_reads+num_al0_h1_long_reads), fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
suppressWarnings(marrangeGrob(list(density.plt, frequency.plt), nrow=2, ncol=1))

ggplot(hdist.filt, aes(num_al0_h0_long_reads/(num_al0_h0_long_reads+num_al0_h1_long_reads)))+geom_density(alpha=0.2)+facet_wrap(~pred_type)
```
## The density plots of each of these bubble groups for number of aligned phased long reads:

```{r}
density.plt <- ggplot(hdist.filt, aes(num_al0_h0_long_reads+num_al0_h1_long_reads, fill=pred_type))+geom_density(alpha=0.2)
frequency.plt <- ggplot(hdist.filt, aes(num_al0_h0_long_reads+num_al0_h1_long_reads, fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
suppressWarnings(marrangeGrob(list(density.plt, frequency.plt), nrow=2, ncol=1))

ggplot(hdist.filt, aes(num_al0_h0_long_reads+num_al0_h1_long_reads))+geom_density(alpha=0.2)+facet_wrap(~pred_type)
```


## The density plots of each of these bubble groups for average kmer coverage:

```{r}
density.plt <- ggplot(hdist.filt, aes(km, fill=pred_type))+geom_density(alpha=0.2)
density.plt

frequency.plt <- ggplot(hdist.filt, aes(km, fill=pred_type))+geom_density(aes(y=..count..), alpha=0.2)
frequency.plt

ggplot(hdist.filt, aes(km))+geom_density(aes(y=..density..))+facet_wrap(~pred_type)
```

Adding threshods for km:

```{r}
quantiles = quantile(hdist.filt[, km], 0.95)
km.hist <- ggplot(hdist.filt, aes(km))+geom_density(aes(y=..density..))

#adding c intercepts:
for (q in quantiles) {
  # adding vertical lines for quantiles
  km.hist = km.hist + geom_vline(data = hdist.filt, xintercept = q, color='red')
}
# adding text annotations
km.hist = km.hist + annotate("text",x=quantiles,y=0.04,label=names(quantiles),hjust=0,color="red", angle = 90, vjust = -0.2)

km.hist
```
## CCS coverage in SS reads

```{r}
ss.cov <- fread('server_mount/SaaRclust/pipeline/ss_cov/ss_cov.data', col.names =  c('cov', 'name'))
ss.cov <- ss.cov[name != 'SSreadNames']
ss.cov.hist <- ggplot(ss.cov, aes(cov))+geom_histogram(binwidth = 1)
quantiles = quantile(ss.cov[, cov], probs = 0.95)
for (q in quantiles) {
  # adding vertical lines for quantiles
  ss.cov.hist <- ss.cov.hist + geom_vline(data = ss.cov, xintercept = q, color='red')
}
ss.cov.hist <- ss.cov.hist + annotate("text",x=quantiles,y=0.04,label=names(quantiles),hjust=0,color="red", angle = 90, vjust = -0.2)

## SS + and - mapping directions per cluster
hard.clust <- get(load('server_mount/SaaRclust/pipeline/aligns_k15_w1_f0.1_z500/SaaRclust_results_HG00733/Clusters/hardClusteringResults_80clusters.RData'))
pb.clust <- data.table(PBreadNames=names(hard.clust$ord), clust=hard.clust$ord)
clust.partners <- findClusterPartners(hard.clust$theta.param)

```



# Conclusion

Adding a set of parameters/thresholds for haplotype clustering:

## For bubbles:
Trim the bubbles based on their kmer coverages. (trim parameter: default=0.05)

## For long reads
1- Add some thresholds on the d1/(d1+d2) value. (default=[0.45, 0.55])

2- Add a paramater for the minimum number of required bubbles (default=3?)